### ISSUE TITLE
[BUG] `debug=True` and `@app.server.before_request` → `AttributeError: '_AppCtxGlobals' object has no attribute 'timing_information'`

### ISSUE THREAD

**Issue by anders-kiaer** [plotly/dash#1475]
**Describe your context**
Please provide us your environment so we can easily reproduce the issue.

-  replace the result of `pip list | grep dash` below
```
dash                                  1.17.0
dash-core-components                  1.13.0
dash-html-components                  1.1.1
dash-renderer                         1.8.3
dash-table                            4.11.0
```

**Describe the bug**

With `debug=True`:

```
Traceback (most recent call last):
  File "[...]/python3.7/site-packages/flask/app.py", line 2464, in __call__
    return self.wsgi_app(environ, start_response)
  File "[...]/python3.7/site-packages/flask/app.py", line 2450, in wsgi_app
    response = self.handle_exception(e)
  File "[...]/python3.7/site-packages/flask/app.py", line 1867, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "[...]/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "[...]/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "[...]/python3.7/site-packages/flask/app.py", line 1953, in full_dispatch_request
    return self.finalize_request(rv)
  File "[...]/python3.7/site-packages/flask/app.py", line 1970, in finalize_request
    response = self.process_response(response)
  File "[...]/python3.7/site-packages/flask/app.py", line 2267, in process_response
    response = handler(response)
  File "[...]/python3.7/site-packages/dash/dash.py", line 1469, in _after_request
    dash_total = flask.g.timing_information["__dash_server"]
  File "[...]/python3.7/site-packages/werkzeug/local.py", line 347, in __getattr__
    return getattr(self._get_current_object(), name)
AttributeError: '_AppCtxGlobals' object has no attribute 'timing_information'
```

**What happens**

In our use case we use the Dash provided Flask server (`app.server`) to set up a `before_request`. On certain conditions (more specifically, certain `GET` parameters in the URL) we do a `flask` redirect. 

In `dash<1.16` the minimal example below works, however in `dash>=1.16` it fails with the above traceback.

**A minimal example**

```python
import flask
import dash
import dash_html_components as html


app = dash.Dash()

app.layout = html.Div("Hello world")

@app.server.before_request
def some_function():
    if flask.request.args.get("redirect_me") == "yes":
        return flask.redirect(flask.request.base_url)
                
if __name__ == "__main__":
    app.run_server(debug=True)
```
Go to http://127.0.0.1:8050/?redirect_me=yes

**Expected behaviour**

Not entirely sure what is best behavior. One possible option: If server timing is not possible to calculate, `return response` without adding header `"Server-Timing"` and displaying it in the frontend (for the given request)?
https://github.com/plotly/dash/blob/ab212a3f90a1d417b4b35de32fb9f6f74e22fc44/dash/dash.py#L1468-L1483

**Comment by alexcjohnson** [comment]
Ah interesting - so when you do a redirect in your own `before_request`, the `before_request` handler we use to initialize timing info is never executed, but our `after_request` handler is? Or perhaps the redirect clears `flask.g`?

Whichever it is, I think it'd be fine to just bail out of our `after_request` if there's no `timing_information` in `flask.g`. In general we only care about these headers for callbacks so it wouldn't matter in your case at all, but even if there were some variant of this in which callbacks lose their timing header that's not a big deal.

**Comment by tophers42** [comment]
Running into this issue. Any recommendations for workarounds?

**Comment by baurt** [comment]
From my understanding, the cause is that the custom `before_request` is causing the Dash `_before_request` function in https://github.com/plotly/dash/blob/ab212a3f90a1d417b4b35de32fb9f6f74e22fc44/dash/dash.py#L1463-L1466 to not be executed.

What I did as a workaround was to create a new `before_request` with the same function that only exists if my app is in debug. I use a `DEBUG` value in my Flask config to denote which mode it should be in. It really shouldn't matter if you check for the mode beforehand, since I would guess that the new `before_request` doesn't add much to response time.  This is a modified version of the above example that works for me:

```python 
import flask
import dash
import dash_html_components as html
import time


app = dash.Dash()
app.server.config["DEBUG"] = True
app.layout = html.Div("Hello world")

if app.server.config["DEBUG"]:
    @app.server.before_request
    def _before_request():
        flask.g.timing_information = {
            "__dash_server": {"dur": time.time(), "desc": None}
        }


@app.server.before_request
def some_function():
    if flask.request.args.get("redirect_me") == "yes":
        return flask.redirect(flask.request.base_url)


if __name__ == "__main__":
    app.run_server(debug=app.server.config.get("DEBUG", False))

```

**Comment by tophers42** [comment]
Thanks @baurt

**Comment by remibaar** [comment]
Issue is resolved in PR #1067 
PR is merged, thus this issue can be closed

**Comment by anders-kiaer** [comment]
@remibaar :tada: :clap: - thanks
